spring-cloud-function-adapters/spring-cloud-function-adapter-azure/src/main/java/org/springframework/cloud/function/adapter/azure/FunctionInvoker.java

```diff
@@ -119,37 +120,117 @@ else if (function == null && StringUtils.hasText(functionDefinition) && APPLICAT
 		return function;
 	}
 
-	@SuppressWarnings({ "unchecked", "rawtypes" })
+	@SuppressWarnings({"unchecked", "rawtypes"})
 	public O handleRequest(I input, ExecutionContext executionContext) {
 		String functionDefinition = executionContext.getFunctionName();
 		FunctionInvocationWrapper function = this.discoverFunction(functionDefinition);
 		Object enhancedInput = enhanceInputIfNecessary(input, executionContext);
 
-		Object output = function.apply(enhancedInput);
-		if (output instanceof Publisher) {
-			if (FunctionTypeUtils.isMono(function.getOutputType())) {
-				return (O) this.convertOutputIfNecessary(input, Mono.from((Publisher) output).blockOptional().get());
+		Object functionResult = function.apply(enhancedInput);
+
+		if (!(functionResult instanceof Publisher)) {
+			return postProcessImperativeFunctionResult(input, enhancedInput, functionResult, function, executionContext);
+		}
+		return postProcessReactiveFunctionResult(input, enhancedInput, (Publisher<?>) functionResult, function, executionContext);
```

Comment:
I think the opposed condition is simpler to understand

```suggestion
		if (functionResult instanceof Publisher) {
			return postProcessReactiveFunctionResult(input, enhancedInput, (Publisher<?>) functionResult, function, executionContext);
		}
		return postProcessImperativeFunctionResult(input, enhancedInput, functionResult, function, executionContext);
```